aiosmtpd/smtp.py

```diff
@@ -74,8 +89,16 @@ def __init__(self, handler,
         self.require_starttls = tls_context and require_starttls
         self._tls_handshake_failed = False
         self._tls_protocol = None
+        self.session = None  # type: Session
+        self.message = None  # type: Envelope
         self.transport = None
 
+    def _create_session(self):
+        return Session()
+
+    def _create_envelope(self):
+        return Envelope()
+
```

Comment:
I have one concern.  With this PR, there are lots of places where we check attributes on the session or envelope objects, and that's fine.  But both attributes start out as None, so if they never get set, then we'll `AttributeError`s.  Ignoring mistakes by subclasses (e.g. overriding either of these methods and accidentally returning None) or evil handlers, I want to be sure that malicious or buggy *clients* can't call SMTP commands out of order to cause these to be None.  It could be an effective DoS attack.

I *think* it's impossible, but I'd like to get your take on it.  Both attributes are effectively initialized when a connection is made.  `connection_made()` calls `_set_rset_state()` which calls `_set_post_data_state()` which sets the envelope.  None of the SMTP commands can be called before the connection is made, so I think that guarantees that the envelope will not be None by then.  `connection_made()` also explicitly calls `_create_session()` so that has to be guaranteed to be initialized too.  And neither attribute ever gets set back to None while the `SMTP` instance lives, so I think we're safe.   Is my reasoning correct?